src/SFML/System/EnumArray.hpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2007-2023 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+#pragma once
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <array>
+
+#include <cassert>
+#include <cstddef>
+
+
+namespace sf::priv
+{
+////////////////////////////////////////////////////////////
+/// \brief Fixed-size array container indexed by an enumeration
+///
+////////////////////////////////////////////////////////////
+template <typename Enum, typename Value, std::size_t Count>
+struct EnumArray : public std::array<Value, Count>
+{
+    ////////////////////////////////////////////////////////////
+    /// \brief Returns a reference to the element associated to specified \a key
+    ///
+    /// No bounds checking is performed in release builds.
+    ///
+    ////////////////////////////////////////////////////////////
+    constexpr Value& operator[](Enum key)
+    {
+        const auto index = static_cast<std::size_t>(key);
+        assert(index < Count && "Index is out of bounds");
+        return std::array<Value, Count>::operator[](index);
+    }
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Returns a reference to the element associated to specified \a key
+    ///
+    /// No bounds checking is performed in release builds.
+    ///
+    ////////////////////////////////////////////////////////////
+    constexpr const Value& operator[](Enum key) const
+    {
+        const auto index = static_cast<std::size_t>(key);
+        assert(index < Count && "Index is out of bounds");
+        return std::array<Value, Count>::operator[](index);
+    }
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Returns a reference to the element associated to specified \a key
+    ///
+    /// \throws std::out_of_range if \a key underlying value is not within array bounds.
+    ///
+    ////////////////////////////////////////////////////////////
+    constexpr Value& at(Enum key)
```

Comment:
Why add `.at()` if these functions are not used internally?

Reply:
I added this as a way of making sure we do not use `at(std::size_t)` for type safety. Should I remove it?

Reply:
Does this prevent the use of `at(std::size_t)`? My impression is that `std::array::at` is still accessible.

Reply:
It accessible with qualified name or casting to base class, like:
```c++
array.std::array<Value, Count>::at(42);
static_cast<const std::array<Value, Count>&>(array).at(42);
```
but not like that:
```c++
array.at(42); // compilation error
```

Reply:
It seems like the type safety benefits are not dependent on these new `at()` overloads existing. If we end up wanting `at()` functions at a later time I'm happy to add them. I just don't want to add code before we need it. Let's go ahead and remove them.